x-pack/legacy/plugins/apm/server/routes/settings.ts

```diff
@@ -4,220 +4,189 @@
-    }
-  });
-
-  // get environments for service
-  server.route({
-    method: 'GET',
-    path: `/api/apm/settings/agent-configuration/services/{serviceName}/environments`,
-    options: {
-      validate: {
-        query: {
-          _debug: Joi.bool()
-        }
-      },
-      tags: ['access:apm']
-    },
-    handler: async req => {
-      const setup = await setupRequest(req);
-      const { serviceName } = req.params;
-      return await getEnvironments({
-        serviceName,
-        setup
-      }).catch(defaultErrorHandler);
-    }
-  });
-
-  const agentConfigPayloadValidation = {
-    settings: Joi.object({
-      transaction_sample_rate: Joi.number()
-        .min(0)
-        .max(1)
-        .precision(3)
-        .required()
-        .options({ convert: false })
+import { minimalRt } from './default_api_types';
+import { createRoute } from './create_route';
+
+// get list of configurations
+export const agentConfigurationRoute = createRoute(core => ({
+  path: '/api/apm/settings/agent-configuration',
+  params: {
+    query: minimalRt
+  },
+  handler: async req => {
+    await createApmAgentConfigurationIndex(core.http.server);
+
+    const setup = await setupRequest(req);
+    return await listConfigurations({
+      setup
+    });
+  }
+}));
+
+// delete configuration
+export const deleteAgentConfigurationRoute = createRoute(() => ({
+  method: 'DELETE',
+  path: '/api/apm/settings/agent-configuration/{configurationId}',
+  params: {
+    path: t.type({
+      configurationId: t.string
+    }),
+    query: minimalRt
+  },
+  handler: async (req, { path }) => {
+    const setup = await setupRequest(req);
+    const { configurationId } = path;
+    return await deleteConfiguration({
+      configurationId,
+      setup
+    });
+  }
+}));
+
+// get list of services
+export const listAgentConfigurationServicesRoute = createRoute(() => ({
+  method: 'GET',
+  path: '/api/apm/settings/agent-configuration/services',
+  params: {
+    query: minimalRt
+  },
+  handler: async req => {
+    const setup = await setupRequest(req);
+    return await getServiceNames({
+      setup
+    });
+  }
+}));
+
+// get environments for service
+export const listAgentConfigurationEnvironmentsRoute = createRoute(() => ({
+  path:
+    '/api/apm/settings/agent-configuration/services/{serviceName}/environments',
+  params: {
+    path: t.type({
+      serviceName: t.string
+    }),
+    query: minimalRt
+  },
+  handler: async (req, { path }) => {
+    const setup = await setupRequest(req);
+    const { serviceName } = path;
+    return await getEnvironments({
+      serviceName,
+      setup
+    });
+  }
+}));
+
+// create configuration
+const transactionSampleRateRt = new t.Type<number, number, unknown>(
+  'TransactionSampleRate',
+  (u): u is number => typeof u === 'number',
+  (u, c) => {
+    return Number(u) >= 0 && Number(u) <= 1
+      ? t.success(Number(u))
+      : t.failure(u, c);
+  },
+  a => Number(a.toPrecision(3))
+);
```

Comment:
Ouch - this just became a whole lot harder to read. Can we stick to joi for this kind of validation or would that not make sense?

Reply:
... I'm not sure `io-ts` is the right tool for the job here, since the type system doesn't care about values - only types. And this is a validation for values.

Reply:
Yeah the joi schema was way nicer. We could use joi, but:
- we'd have to define the io-ts type anyway
- we'd need to add custom logic to throw based on the joi validation
- we can't use joi in the browser (for now). this is why we have separate client-side validation logic, see x-pack/legacy/plugins/apm/public/components/app/Settings/AddSettings/AddSettingFlyout.tsx. with io-ts we could remove that duplication.

what would you prefer? we can also create a generic range type for io-ts, and use that. something like:
`const transactionSampleRateRt = range(1,3)`.